authorization/authorization_integration_test.go

```diff
@@ -60,837 +60,272 @@ var _ = Describe("Authorized", func() {
-								"Actions": MatchAllKeys(Keys{
-									action: PointTo(MatchFields(IgnoreExtras, Fields{
-										"Allow": Equal(false),
-									})),
-								}),
-							})),
-						}),
-					})),
-				}),
-			})))
-		})
-
-		It("IsAuthorizedDTResourceNonValid", func() {
-			var err error
-			authorizationClient, err := integration.InitConfigAuthorization()
-			Expect(err).To(Succeed())
-
-			digitalTwin := &authorizationpb.DigitalTwin{
-				Id: integration.Node1,
-			}
-
-			resources := integration.Resource2
-			// To make sure that the proper audit log was queried from BigQuery, need to add a unique identifier.
-			noAuditLogEntry = true
-			inputParams := map[string]*authorizationpb.InputParam{
-				"auditLog": {
-					Value: &authorizationpb.InputParam_StringValue{
-						StringValue: fmt.Sprintf("\"%v\"", auditLogIdentifier),
+				}
+
+				resp, err := authorizationClient.IsAuthorized(
+					context.Background(),
+					digitalTwin,
+					resources,
+					inputParams,
+					policyTags,
+					retry.WithMax(5),
+				)
+
+				if expectedError != "" {
+					if numberResources == 0 {
+						noAuditLogEntry = true
+					}
+					Expect(err).To(MatchError(ContainSubstring(expectedError)))
+					Expect(resp).To(BeNil())
+				} else {
+					Expect(err).To(Succeed())
+					Expect(resp).NotTo(BeNil())
+
+					decision := resources[0].Type
+					resource := resources[0].ExternalId
+					action := resources[0].Actions[0]
+
+					if numberResources == 2 {
+						resource1 := resources[1].ExternalId
+						action1 := resources[1].Actions[0]
+						Expect(resp).To(PointTo(MatchFields(IgnoreExtras, Fields{
+							"DecisionTime": Not(BeNil()),
+							"Decisions": MatchAllKeys(Keys{
+								decision: PointTo(MatchFields(IgnoreExtras, Fields{
+									"Resources": MatchAllKeys(Keys{
+										resource: PointTo(MatchFields(IgnoreExtras, Fields{
+											"Actions": MatchAllKeys(Keys{
+												action: PointTo(MatchFields(IgnoreExtras, Fields{
+													"Allow": Equal(true),
+												})),
+											}),
+										})),
+										resource1: PointTo(MatchFields(IgnoreExtras, Fields{
+											"Actions": MatchAllKeys(Keys{
+												action1: PointTo(MatchFields(IgnoreExtras, Fields{
+													"Allow": Equal(true),
+												})),
+											}),
+										})),
+									}),
+								})),
+							}),
+						})))
+					} else {
+						Expect(resp).To(PointTo(MatchFields(IgnoreExtras, Fields{
+							"DecisionTime": Not(BeNil()),
+							"Decisions": MatchAllKeys(Keys{
+								decision: PointTo(MatchFields(IgnoreExtras, Fields{
+									"Resources": MatchAllKeys(Keys{
+										resource: PointTo(MatchFields(IgnoreExtras, Fields{
+											"Actions": MatchAllKeys(Keys{
+												action: PointTo(MatchFields(IgnoreExtras, Fields{
+													"Allow": Equal(expectedAllow),
+												})),
+											}),
+										})),
+									}),
+								})),
+							}),
+						})))
+					}
```

Comment:
you could rewrite this using some helper function like e.g. (didn't try if it actually works)):
```Go
func matchDecisions(resources []*authorizationpb.IsAuthorizedRequest_Resource, allow bool) types.GomegaMatcher {
    typeToResourceKeys := map[any]Keys
    for _, resource := range resources {
        resourceKeys, ok := typeToResourceKeys[resource.Type]
        if !ok {
            resourceKeys = Keys{}
            typeToResourceKeys[resource.Type] = resourceKeys
        }
        actionKeys := Keys{}
        for _, action := range resource.Actions {
            actionsKeys[action] = PointTo(MatchFields(IgnoreExtras, Fields{
                "Allow": Equal(allow),
            }))
        }
        resourceKeys[resource.ExternalId] = PointTo(MatchFields(IgnoreExtras, Fields{
            "Actions": MatchAllKeys(actionKeys),
        }))
    }
    decisionKeys := Keys{}
    for resourceType, resourceKeys := range typeToResources {
        decisionKeys[resourceType] = PointTo(MatchFields(IgnoreExtras, Fields{
            "Resources": MatchAllKeys(resourceKeys),
        }))
    }
    return MatchAllKeys(decisionKeys)
}
```

But I'll let you decide if it is the way to write tests - one could argue it is too complex for being a test code :)

Reply:
And now I see this "clever" solution may be even more code than your solution :)

Reply:
Please ignore this suggestion above and rather split these test cases, so there is no need for `if numberResources == 2` condition...